src/graph/renderer/CytoscapeGraphLayoutAdapter.ts

```diff
@@ -0,0 +1,59 @@
+import { Ext, LayoutManipulation, LayoutPositionOptions, Core } from 'cytoscape'
+import { GraphModel } from '../GraphModel'
+import { GraphLayout, GraphLayoutAlgorithm } from '../types'
+
+export type CustomLayoutOptions = {
+  model: GraphModel
+  algorithm: GraphLayoutAlgorithm
+}
+
+export class CytoscapeGraphLayoutAdapter implements LayoutManipulation {
+  private readonly options: LayoutPositionOptions & CustomLayoutOptions
+
+  static readonly LAYOUT_NAME: GraphLayout = 'custom'
+
+  constructor(options: LayoutPositionOptions & CustomLayoutOptions) {
+    this.options = options
+  }
+
+  static register: Ext = (cytoscape) => {
+    if (cytoscape == null) {
+      return
+    } // can't register if cytoscape unspecified
+    cytoscape(
+      'layout',
+      // @ts-ignore
+      CytoscapeGraphLayoutAdapter.LAYOUT_NAME,
+      // @ts-ignore
+      CytoscapeGraphLayoutAdapter
+    )
+  }
+
+  run(): this {
+    if (this.options.algorithm == null) {
+      throw new Error('No custom layout algorithm was specified')
+    }
+
+    //@ts-ignore
+    const cy: Core = this.options.cy
+    const boundingBox = { x1: 0, y1: 0, w: cy.width(), h: cy.height() }
+    const nodePositions = this.options.algorithm.runLayout(this.options.model, {
```

Comment:
Q: Is there a way to allow the runLayout to be async?

I'm thinking that complex layout algorithms could be run in a web-worker and update the positions asynchronously.